File: backend/src/main/kotlin/com/example/backend/controllers/CommandeController.kt
package com.example.backend.controllers

import com.example.backend.dtos.*
import com.example.backend.models.Commande
import com.example.backend.services.CommandeService
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*
import java.io.File

@RestController
@RequestMapping("/api/commandes")
class CommandeController(
  private val commandeService: CommandeService
) {


  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping
  fun creerCommande(@RequestBody commandeDto: CommandeRequest): ResponseEntity<Commande> {
    val commande = commandeService.createCommande(commandeDto)
    return ResponseEntity.ok(commande)
  }


  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/reception")
  fun receptionnerCommande(
    @PathVariable id: Long,
    @RequestParam receptionType: String,
    @RequestBody productCmdList: List<ProduitCmdRequest>
  ): ResponseEntity<Commande> {
    val commande = commandeService.receptionnerCommande(id, receptionType, productCmdList)
    return ResponseEntity.ok(commande)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping
  fun getAllCommandesMapped(): ResponseEntity<List<Map<String, Any?>>> {
    val commandes = commandeService.getAllCommandesMapped()
    return ResponseEntity.ok(commandes)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/paged")
  fun getAllCommandesMapped(
    @RequestParam(defaultValue = "0") page: String,
    @RequestParam(defaultValue = "10") size: String,
    @RequestParam(defaultValue = "id") sortBy: String,
    @RequestParam(required = false) search: String?,
    @RequestParam(required = false) etat: String?,
    @RequestParam(required = false) typeFournisseur: String?,
    @RequestParam(required = false) fournisseurId: String?,
    @RequestParam(required = false) startDate: String?,
    @RequestParam(required = false) endDate: String?
  ): ResponseEntity<CommandePageableCustomlDto> {
    val pageNumber = page.toIntOrNull()?.coerceAtLeast(0) ?: 0
    val pageSize = size.toIntOrNull()?.coerceAtLeast(1) ?: 10
    val pageable = PageRequest.of(pageNumber, pageSize, Sort.by(Sort.Direction.DESC, "dateCreation"))
    val commandes =
      commandeService.getAllCommandesMappedPageable(pageable, etat, fournisseurId, typeFournisseur, startDate, endDate)
    return ResponseEntity.ok(commandes)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{id}")
  fun obtenirCommande(@PathVariable id: Long): ResponseEntity<Map<String, Any?>> {
    val commande = commandeService.getCommandeById(id)
    return ResponseEntity.ok(commande)
  }


  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/modifier-lignes")
  fun modifierLignes(@PathVariable id: Long, @RequestBody produits: List<ProduitCmdRequest>): ResponseEntity<Commande> {
    val commande = commandeService.modifierLignesCommande(id, produits)
    return ResponseEntity.ok(commande)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/commande_par_fournisseur")
  fun commandeByFournisseur(
    @RequestParam fournisseurId: String?,
    @RequestParam totalAmount: String,
    @RequestBody produits: List<CommandeNewDTO>
  ): ResponseEntity<Commande> {
    val commande = commandeService.commandeByFournisseur(fournisseurId, totalAmount, produits)
    return ResponseEntity.ok(commande)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @DeleteMapping("/{id}")
  fun supprimerCommande(@PathVariable id: Long): ResponseEntity<Void> {
    commandeService.supprimerCommande(id)
    return ResponseEntity.noContent().build()
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/ajouter-fournisseur")
  fun ajouterFournisseur(@PathVariable id: Long, @RequestParam fournisseurId: Long): ResponseEntity<Commande> {
    val commande = commandeService.ajouterFournisseur(id, fournisseurId)
    return ResponseEntity.ok(commande)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/annuler")
  fun annulerCommande(@PathVariable id: Long): ResponseEntity<Commande> {
    val commande = commandeService.annulerCommande(id)
    return ResponseEntity.ok(commande)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/imprimer-bon-pdf")
  fun imprimerBonPdf(@PathVariable id: Long): ResponseEntity<ByteArray> {
    val outputPath = "temp_bon_commande.pdf" // Temporary file path
    commandeService.imprimerBonPdf(id, outputPath)

    val file = File(outputPath)
    val fileContent = file.readBytes()
    file.delete() // Clean up the temporary file

    return ResponseEntity.ok()
      .header("Content-Disposition", "attachment; filename=bon_commande.pdf")
      .contentType(org.springframework.http.MediaType.APPLICATION_PDF)
      .body(fileContent)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/reception-complementaire")
  fun receptionComplementaire(
    @PathVariable id: Long,
    @RequestBody produits: List<ProduitCmdRequest>
  ): ResponseEntity<Commande> {
    val commande = commandeService.receptionComplementaire(id, produits)
    return ResponseEntity.ok(commande)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/ajouter-justificatif")
  fun ajouterJustificatif(@PathVariable id: Long, @RequestParam justificatif: String): ResponseEntity<Commande> {
    val commande = commandeService.ajouterJustificatif(id, justificatif)
    return ResponseEntity.ok(commande)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{id}/historique-reception")
  fun visualiserHistorique(@PathVariable id: Long): ResponseEntity<List<Map<String, Any?>>> {
    val historique = commandeService.visualiserHistoriqueReception(id)
    return ResponseEntity.ok(historique)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/cloturer")
  fun cloturerCommande(@PathVariable id: Long): ResponseEntity<Map<String, Any?>> {
    val commande = commandeService.cloturerCommande(id)
    return ResponseEntity.ok(commande)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/ajouter-facture")
  fun ajouterFacture(@PathVariable id: Long, @RequestParam facture: String): ResponseEntity<Commande> {
    val commande = commandeService.ajouterFacture(id, facture)
    return ResponseEntity.ok(commande)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/generer-rapport")
  fun genererRapport(@PathVariable id: Long): ResponseEntity<String> {
    var data = commandeService.genererRapportLivraison(id)
    return ResponseEntity.ok(data)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{id}/exporter")
  fun exporterCommande(@PathVariable id: Long): ResponseEntity<String> {
    var data = commandeService.exporterCommande(id)
    return ResponseEntity.ok(data)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/{id}/ajouter-motif-annulation")
  fun ajouterMotifAnnulation(@PathVariable id: Long, @RequestParam motif: String): ResponseEntity<Void> {
    commandeService.annulerCommande(id)
    commandeService.ajouterMotifAnnulation(id, motif)
    return ResponseEntity.noContent().build()
  }

//  @PutMapping("/{id}")
//  fun validerCommande(@PathVariable id: Long): ResponseEntity<Commande> {
//    val commande = commandeService.validerCommande(id)
//    return ResponseEntity.ok(commande)
//  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/product/paged")
  fun getCommandeInfoByProduct(
    @RequestParam(defaultValue = "0") page: String,
    @RequestParam(defaultValue = "10") size: String,
    @RequestParam(defaultValue = "id") sortBy: String,
    @RequestParam(required = false) produitId: String?,
    @RequestParam(required = false) startDate: String?,
    @RequestParam(required = false) endDate: String?
  ): ResponseEntity<CommandePageableCustomlDto> {
    val pageNumber = page.toIntOrNull()?.coerceAtLeast(0) ?: 0
    val pageSize = size.toIntOrNull()?.coerceAtLeast(1) ?: 10
    val pageable = PageRequest.of(pageNumber, pageSize, Sort.by(Sort.Direction.DESC, "commande.dateCreation"))
    val commandes = commandeService.getCommandeInfoByProduct(pageable, produitId, startDate, endDate)
    return ResponseEntity.ok(commandes)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/product-commande/update")
  fun updateCommandeSimple(
    @RequestParam(required = false) commandeId: String?,
    @RequestParam(required = false) produitCmdId: String?,
    @RequestParam(required = false) qteRecu: String?,
    @RequestParam(required = false) prixAchat: String?,
    @RequestParam(required = false) prixVente: String?,
  ): ResponseEntity<Commande> {
    val commandes = commandeService.updateCommandeSimple(
      commandeId,
      produitCmdId,
      qteRecu,
      prixAchat,
      prixVente
    )
    return ResponseEntity.ok(commandes)
  }
}

File: backend_py/pharmaPython/app/api/v1/commande_controller.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional, Dict, Any

from app.api.deps import get_db

from app.models.commandeout import CommandeSchema
from app.models.employe import Employe
from app.schemas.commande_dto import ProduitCmdRequest, CommandeNewDTO, CommandePageableCustomlDto, CommandeRequest, \
  CommandeRuptureRequest
from app.services.commande_service import CommandeService
from app.utility.jwt_authentication import jwt_authentication
from app.utility.user_utils import UserUtils

router = APIRouter(
  prefix="/commandes",
  tags=["Commandes"],
  dependencies=[Depends(jwt_authentication)],
)


# ----------------------------
# Créer une commande
# ----------------------------
@router.post("/", response_model=CommandeSchema)
def creer_commande(commande_dto: CommandeRequest, db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.create_commande(commande_dto)


# ----------------------------
# Réceptionner une commande
# ----------------------------
@router.post("/{id}/reception", response_model=CommandeSchema)
def receptionner_commande(
  id: int,
  receptionType: str,
  produits: List[ProduitCmdRequest],
  db: Session = Depends(get_db)
):
  service = CommandeService(db)
  return service.receptionner_commande(id, receptionType, produits)


# ----------------------------
# Liste des commandes (mappées)
# ----------------------------
@router.get("/", response_model=List[Dict[str, Any]])
def get_all_commandes_mapped(db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.get_all_commandes_mapped()


# ----------------------------
# Liste des commandes paginées
# ----------------------------
@router.get("/paged")
def get_all_commandes_paged(
  page: int = Query(0, ge=0),
  size: int = Query(10, ge=1),
  etat: Optional[str] = None,
  typeFournisseur: Optional[str] = None,
  fournisseurId: Optional[str] = None,
  startDate: Optional[str] = None,
  endDate: Optional[str] = None,
  db: Session = Depends(get_db)
):
  service = CommandeService(db)
  return service.get_all_commandes_mapped_pageable(page, size, etat, fournisseurId, typeFournisseur, startDate, endDate)


# ----------------------------
# Obtenir une commande par ID
# ----------------------------
@router.get("/{id}", response_model=Dict[str, Any])
def obtenir_commande(id: int, db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.get_commande_by_id(id)


# ----------------------------
# Modifier lignes d'une commande
# ----------------------------
@router.post("/{id}/modifier-lignes", response_model=CommandeSchema)
def modifier_lignes(id: int, produits: List[ProduitCmdRequest], db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.modifier_lignes_commande(id, produits)


# ----------------------------
# Créer une commande fournisseur
# ----------------------------
@router.post("/commande_par_fournisseur")
def commande_by_fournisseur(
  fournisseurId: Optional[str],
  totalAmount: str,
  produits: List[CommandeNewDTO],
  db: Session = Depends(get_db)
):
  service = CommandeService(db)
  return service.commande_by_fournisseur(fournisseurId, totalAmount, produits)


@router.post("/commande_par_fournisseur/rupture")
def commande_by_fournisseur(
  fournisseurId: Optional[str],
  totalAmount: str,
  produits: List[CommandeNewDTO],
  db: Session = Depends(get_db)
):
  service = CommandeService(db)
  return service.commande_by_fournisseur(fournisseurId, totalAmount, produits)


# ----------------------------
# Supprimer une commande
# ----------------------------
@router.delete("/{id}", status_code=204)
def supprimer_commande(id: int, db: Session = Depends(get_db)):
  service = CommandeService(db)
  service.supprimer_commande(id)
  return {"message": "Commande supprimée"}


# ----------------------------
# Ajouter fournisseur à une commande
# ----------------------------
@router.post("/{id}/ajouter-fournisseur", response_model=CommandeSchema)
def ajouter_fournisseur(id: int, fournisseurId: int, db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.ajouter_fournisseur(id, fournisseurId)


# ----------------------------
# Annuler une commande
# ----------------------------
@router.post("/{id}/annuler", response_model=CommandeSchema)
def annuler_commande(id: int, db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.annuler_commande(id)


# ----------------------------
# Imprimer bon PDF
# ----------------------------
@router.post("/{id}/imprimer-bon-pdf")
def imprimer_bon_pdf(id: int, db: Session = Depends(get_db)):
  service = CommandeService(db)
  pdf_bytes = service.imprimer_bon_pdf(id)
  return {
    "filename": "bon_commande.pdf",
    "content": pdf_bytes
  }


# ----------------------------
# Réception complémentaire
# ----------------------------
@router.post("/{id}/reception-complementaire", response_model=CommandeSchema)
def reception_complementaire(id: int, produits: List[ProduitCmdRequest], db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.reception_complementaire(id, produits)


# ----------------------------
# Ajouter justificatif
# ----------------------------
@router.post("/{id}/ajouter-justificatif", response_model=CommandeSchema)
def ajouter_justificatif(id: int, justificatif: str, db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.ajouter_justificatif(id, justificatif)


# ----------------------------
# Historique de réception
# ----------------------------
@router.get("/{id}/historique-reception", response_model=List[Dict[str, Any]])
def visualiser_historique(id: int, db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.visualiser_historique_reception(id)


# ----------------------------
# Clôturer commande
# ----------------------------
@router.post("/{id}/cloturer", response_model=Dict[str, Any])
def cloturer_commande(id: int, db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.cloturer_commande(id)


# ----------------------------
# Ajouter facture
# ----------------------------
@router.post("/{id}/ajouter-facture", response_model=CommandeSchema)
def ajouter_facture(id: int, facture: str, db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.ajouter_facture(id, facture)


# ----------------------------
# Générer rapport livraison
# ----------------------------
@router.post("/{id}/generer-rapport", response_model=str)
def generer_rapport(id: int, db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.generer_rapport_livraison(id)


# ----------------------------
# Exporter commande
# ----------------------------
@router.get("/{id}/exporter", response_model=str)
def exporter_commande(id: int, db: Session = Depends(get_db)):
  service = CommandeService(db)
  return service.exporter_commande(id)


# ----------------------------
# Ajouter motif d’annulation
# ----------------------------
@router.post("/{id}/ajouter-motif-annulation", status_code=204)
def ajouter_motif_annulation(id: int, motif: str, db: Session = Depends(get_db)):
  service = CommandeService(db)
  service.annuler_commande(id)
  service.ajouter_motif_annulation(id, motif)
  return {"message": "Motif ajouté"}


# ----------------------------
# Commande info par produit (paged)
# ----------------------------
@router.get("/product/paged", response_model=CommandePageableCustomlDto)
def get_commande_info_by_product(
  page: int = Query(0, ge=0),
  size: int = Query(10, ge=1),
  produitId: Optional[str] = None,
  startDate: Optional[str] = None,
  endDate: Optional[str] = None,
  db: Session = Depends(get_db)
):
  service = CommandeService(db)
  return service.get_commande_info_by_product(page, size, produitId, startDate, endDate)


# ----------------------------
# Update commande simple (produit)
# ----------------------------
@router.post("/product-commande/update", response_model=CommandeSchema)
def update_commande_simple(
  commandeId: Optional[str] = None,
  produitCmdId: Optional[str] = None,
  qteRecu: Optional[str] = None,
  prixAchat: Optional[str] = None,
  prixVente: Optional[str] = None,
  db: Session = Depends(get_db)
):
  service = CommandeService(db)
  return service.update_commande_simple(commandeId, produitCmdId, qteRecu, prixAchat, prixVente)

@router.post("/product-commande/rupture")
def commande_rupture(commande_dto: CommandeRuptureRequest, db: Session = Depends(get_db), employe: Employe = Depends(UserUtils.get_current_employe)):
  service = CommandeService(db)
  return service.reapprovisionner_rupture(commande_dto,employe)
